src/CMakeLists.txt

```diff
@@ -375,6 +395,59 @@ if (ENABLE_BLACKMAGIC)
   endif()
 endif()
 
+################## OPENCV ###################
+
+find_package( OpenCV 4 )
+if (OpenCV_FOUND)
+  message("\nCOMPILING WITH OPENCV\n")
+  set(CMAKE_SWIG_FLAGS "-DUSE_OPENCV=1")
+  add_definitions( -DUSE_OPENCV=1 )
+else()
+  message("\nOPENCV NOT FOUND, SOME FUNCTIONALITIES WILL BE DISABLED\n")
+endif()
+
+################## PROTOBUF ##################
+if (OpenCV_FOUND)
+  find_package(Protobuf 3)
+  
+  if (NOT Protobuf_FOUND)
+    # Protobuf is required when compiling with opencv
+    message(FATAL_ERROR "\nPLEASE INSTALL PROTOBUF. Protobuf is required when compiling with opencv.\n")
+  endif()
+  
+  # compiling protobuf messages
+  include_directories(${PROTOBUF_INCLUDE_DIR})
+  set(PROTOBUF_MESSAGES_FOLDER "${CMAKE_CURRENT_SOURCE_DIR}/protobuf_messages")
+
+  FILE(GLOB PROTO_MSGS "${PROTOBUF_MESSAGES_FOLDER}/*.proto")
+  foreach(PROTO_MSG ${PROTO_MSGS})
+    execute_process (
+      COMMAND bash -c "${PROTOBUF_PROTOC_EXECUTABLE} -I=${PROTOBUF_MESSAGES_FOLDER} --cpp_out=${PROTOBUF_MESSAGES_FOLDER} ${PROTO_MSG}"
+    )
+  endforeach()
+
+  FILE(GLOB PROTO_CCS "${PROTOBUF_MESSAGES_FOLDER}/*.pb.cc")
+  foreach(PROTO_CC ${PROTO_CCS})
+    execute_process(COMMAND ${CMAKE_COMMAND} -E copy ${PROTO_CC} ${CMAKE_CURRENT_SOURCE_DIR})
+    file(REMOVE ${PROTO_CC})
+  endforeach()
+  
+  FILE(GLOB PROTO_HS "${PROTOBUF_MESSAGES_FOLDER}/*.pb.h")
+  foreach(PROTO_H ${PROTO_HS})
+    execute_process(COMMAND ${CMAKE_COMMAND} -E copy ${PROTO_H} ${CMAKE_CURRENT_SOURCE_DIR})
+    file(REMOVE ${PROTO_H})
+  endforeach()
```

Comment:
Modifying files in the _source_ dir is **not** OK during a build.

The CMake Find module for Protobuf has an example of how to use the macros it defines to generate source files:
(From `/usr/share/cmake/Modules/FindProtobuf.cmake` on my machine...)
```cmake
  find_package(Protobuf REQUIRED)
  include_directories(${Protobuf_INCLUDE_DIRS})
  include_directories(${CMAKE_CURRENT_BINARY_DIR})
  protobuf_generate_cpp(PROTO_SRCS PROTO_HDRS foo.proto)
  protobuf_generate_cpp(PROTO_SRCS PROTO_HDRS EXPORT_MACRO DLL_EXPORT foo.proto)
  protobuf_generate_cpp(PROTO_SRCS PROTO_HDRS DESCRIPTORS PROTO_DESCS foo.proto)
  protobuf_generate_python(PROTO_PY foo.proto)
  add_executable(bar bar.cc ${PROTO_SRCS} ${PROTO_HDRS})
  target_link_libraries(bar ${Protobuf_LIBRARIES})

.. note::
  The ``protobuf_generate_cpp`` and ``protobuf_generate_python``
  functions and :command:`add_executable` or :command:`add_library`
  calls only work properly within the same directory.
```
You don't actually need the third line as that's already automatic, and obviously the last two lines should be replaced with configuration on the library instead. I'd also prefer `target_include_directories(openshot PUBLIC ${Protobuf_INCLUDE_DIRS})` to the directory-wide `include_dirs()`.

Using their macros will generate files into the _build_ directory and reference them from there. It'll also add the files to CMake's list of `GENERATED` files to delete during a `make clean`. If the example isn't adaptable to what you need to do, let me know what the issues are and we can figure something out that still maintains the immutability of the source tree.

I'll most likely still want to encapsulate all of this into an `INTERFACE` target, that way it'll automatically be carried with the `openshot` target across directories (you won't have to do anything in `tests/` but add the test source files), and it'll become part of the `EXPORTED` library configs automatically. But I can tackle that part later.

Reply:
Yeah this was far from ideal. At the time I was having problems including the message headers when I compiled the messages on build folder. I'll take another look at this and follow your suggestion. Thanks.

Reply:
@BrennoCaldato 

Anytime you add include files to a build, you need to jump through a bunch of hoops. Doubly so if they become public headers, which of course they do if they get referenced in any _other_ public headers. And if those includes are **generated** files, then it gets really complex. 

1. There needs to be a step to generate the files, which has to be done before the build step for anything that includes them. (If the files can be generated _during_ the configure run, that's a lot easier because it eliminates the need to add build-step dependencies.)

1. The generated files need to be installed after the build is done. Which means you need to create a list of all of the files that _will be_ generated, and where they _will be_ on the disk, so the correct install scripts can be generated. You can't (ever) glob for _generated_ files in `CMakeLists.txt` scripting, because those output files **don't exist yet**. 

    In CMake's processing, generated files get produced _after_ the configuration is generated. CMake processes all of the `CMakeLists.txt` files, writes out the control files and scripting for everything specified at the various levels, and **then** executes the file-generation scripts. So at the time the glob is actually performed, there's nothing to find. (That's true whether the files in question are in the source tree or the build tree, and it's why there's never any reason (or need) to generate files into the build tree — it won't change the situation in any way that matters. Using build-tree files during the build is just a matter of adding the right configuration, same as source-tree files.)

1. Because new headers are being created and installed, their _location_ changes between the build context and the install context, so any references to them are contextual. The location of the install tree can also change, so _that_ context also **must** use only relative paths so they can be rooted at an as-yet-undetermined `CMAKE_INSTALL_PREFIX`. Otherwise, you end up with a build that succeeds, but then installs broken headers or configuration.

It can be a pain in the ass to get right, but that's why CMake includes a set of macros in their Protobuf configuration (the Find module) that should take care of most of the tedium automatically.

Reply:
> and it's why there's never any reason (or need) to generate files into the build tree

Er, I meant into the \*source tree.

Reply:
One of the reasons Kitware (the CMake developers) recommend so strongly against globbing for files _period_ in `CMakeLists.txt` scripts is, it's easy to trick yourself into thinking something works, and sending yourself down wrong paths based on that apparent success.

If a `CMakeLists.txt` script uses `file(GLOB...` commands, it'll work fine if at some point you managed to get the necessary files generated beforehand. Maybe a previous run partly succeeded, maybe you generated them by hand, whatever. Regardless how, they're there, so now CMake has something to find when it's _**re**-generating_ the configuration. It'll succeed as long as it's run in in a "dirty" tree. But _only_ in a dirty tree. Later, you go to test your configuration in a _clean_ tree, and surprise! Suddenly it's failing again. So then you start adding more complexity to try to "fix" the globs.

Stop me if any of this is sounding familiar. :wink:  Been there, done that, plenty of times. Finally learned my lesson, but only after far too much time wandering down those wrong paths.

The reality is that there's no way to "fix" the globs no matter how hard we try. They can't ever work that way, and the only real fix is to find a way to do things that doesn't require them.

Reply:
@ferdnyc 

I forgot to answer this message, sorry. 
The protobuf generated files issue was fixed some commits ago. I created a library for the proto messages and the files are built outside of src now.

Thanks again for the detailed explanation. Aways very informative.